optuna/visualization/_optimization_history.py

```diff
@@ -23,6 +24,95 @@
 _logger = get_logger(__name__)
 
 
+class _ValuesInfo(NamedTuple):
+    values: List[float]
+    stds: Optional[List[float]]
+    label_name: str
+
+
+class _OptimizationHistoryInfo(NamedTuple):
+    trial_numbers: List[int]
+    values_info: _ValuesInfo
+    best_values_info: Optional[_ValuesInfo]
+
+
+def _get_optimization_history_info_list(
+    study: Union[Study, Sequence[Study]],
+    target: Optional[Callable[[FrozenTrial], float]],
+    target_name: str,
+    error_bar: bool,
+) -> List[_OptimizationHistoryInfo]:
+
+    _check_plot_args(study, target, target_name)
+    if isinstance(study, Study):
+        studies = [study]
+    else:
+        studies = list(study)
+
+    info_list: List[_OptimizationHistoryInfo] = []
+    for study in studies:
+        trials = study.get_trials(states=(TrialState.COMPLETE,))
+        label_name = target_name if len(studies) == 1 else f"{target_name} of {study.study_name}"
+        if target is not None:
+            values = [target(t) for t in trials]
+            # We don't calculate best for user-defined target function since we cannot tell
+            # which direction is better.
+            best_values_info: Optional[_ValuesInfo] = None
+        else:
+            values = [cast(float, t.value) for t in trials]
+            if study.direction == StudyDirection.MINIMIZE:
+                best_values = list(np.minimum.accumulate(values))
+            else:
+                best_values = list(np.maximum.accumulate(values))
+            best_label_name = (
+                "Best Value" if len(studies) == 1 else f"Best Value of {study.study_name}"
+            )
+            best_values_info = _ValuesInfo(best_values, None, best_label_name)
+        info_list.append(
+            _OptimizationHistoryInfo(
+                trial_numbers=[t.number for t in trials],
+                values_info=_ValuesInfo(values, None, label_name),
+                best_values_info=best_values_info,
+            )
+        )
+
+    if len(info_list) == 0:
+        _logger.warning("There are no studies.")
+
+    if sum(len(info.trial_numbers) for info in info_list) == 0:
+        _logger.warning("There are no complete trials.")
+        info_list.clear()
```

Comment:
Currently, `info_list` contains n `_OptimizationHistoryInfo`s when there are n studies, except when all studies are empty.
Is there some reason for this (seemingly inconsistent) behavior?

Reply:
I followed the [original code](https://github.com/optuna/optuna/blob/191f4db7228eb205e409f9c44fa1f64595823194/optuna/visualization/_optimization_history.py#L110-L112).
I don't know why, but it is tested [here](https://github.com/optuna/optuna/blob/191f4db7228eb205e409f9c44fa1f64595823194/tests/visualization_tests/test_optimization_history.py#L146-L148), so it throws an error if we return a list of n empty lists instead of a single empty list.

Reply:
I see. I think fixing this behavior can be another PR.